### PR TITLE
[Perf] Auto-compile trivial CustomOp fallbacks to complete GemmaRMSNorm fusion under enforce_eager

### PR DESCRIPTION
**Pull request by Ray-RP**
# [Perf] Auto-compile trivial CustomOp fallbacks to complete GemmaRMSNorm fusion under enforce_eager

## Background

v0.20.0 added a `vllm_c` dispatch path for `RMSNorm`, which fuses the standard norm kernel under `enforce_eager=True`. This PR extends that fix to the classes that were missed: **`GemmaRMSNorm`, `Gemma2RMSNorm`, and similar ops that still delegate `forward_cuda` to an unfused `forward_native`**.

## Problem

`GemmaRMSNorm` (used by all Gemma-family models including Gemma3) has:

```python
class GemmaRMSNorm(CustomOp):
    def forward_cuda(self, x, residual=None):
        return self.forward_native(x, residual)   # trivial fallback — no CUDA kernel
    
    def forward_native(self, x, residual=None):
        # pure PyTorch: weight + 1 scaling variant
        ...
```

Unlike `RMSNorm`, which now routes through `vllm_c`, `GemmaRMSNorm.forward_cuda` is a one-liner alias. Every call under `enforce_eager=True` produces **8 unfused elementwise/reduce kernels** instead of 1 fused Triton kernel:

```
aten::mul, aten::pow, aten::mean, aten::add, aten::rsqrt, aten::mul, aten::mul, aten::add
```

We audited all `CustomOp` subclasses in v0.20.0 and found **14 subclasses** still exhibit this pattern after the `RMSNorm` fix:

| Class | Unfused kernels |
|-------|----------------|
| `GemmaRMSNorm` | 8 |
| `Gemma2RMSNorm` | 8 |
| `RMSNormWithoutScale` | 6 |
| `ReLUSquaredActivation` | 3 |
| … 10 more | 3–8 |

These collectively account for **~40% of GPU kernel time** in Gemma3-4B-IT under `enforce_eager=True`.

## Why enforce_eager Matters

Users land in `enforce_eager=True` via:
- **Quantization**: BitsAndBytes 8-bit auto-sets it (`config/model.py:1057-1070`)
- **ROCm + encoder-decoder**: T5/BART/Whisper on AMD GPU (`config/model.py:1036-1042`)
- **DeepSeek V3 + MTP speculative decoding** (`config/speculative.py:465-468`, marked FIXME)
- **`TORCH_COMPILE_DISABLE=1`** env var, debugging, LoRA multi-adapter, OOM fallback

The fragmentation is **invisible**: GPU utilization stays high, `torch.profiler` under CUDAGraph replay hides it entirely.

## Proposed Fix

Detect the trivial-fallback pattern at subclass registration time (`__init_subclass__`) and replace `forward_cuda` with a `torch.compile`'d version when `VLLM_AUTO_FUSE_OPS=1`:

```python
# vllm/model_executor/custom_op.py

import inspect, os

def _is_trivial_forward_cuda(cls) -> bool:
    """Return True if forward_cuda is a one-liner delegating to forward_native."""
    try:
        src = inspect.getsource(cls.forward_cuda)
    except (OSError, TypeError):
        return False
    body = [
        l.strip() for l in src.splitlines()
        if l.strip()
        and not l.strip().startswith(("#", "@", "def ", '"""', "'''"))
    ]
    return len(body) == 1 and body[0].startswith("return") and "forward_native" in body[0]


class CustomOp(torch.nn.Module):
    def __init_subclass__(cls, **kwargs):
        super().__init_subclass__(**kwargs)
        # ... existing registration logic ...

        if (
            os.environ.get("VLLM_AUTO_FUSE_OPS") == "1"
            and _is_trivial_forward_cuda(cls)
            and not getattr(cls, "_auto_fused", False)
        ):
            _original_native = cls.forward_native
            _compiled = torch.compile(_original_native, dynamic=True, mode="default")
            def _fused_forward_cuda(self, *args, **kwargs):
                return _compiled(self, *args, **kwargs)
            cls.forward_cuda = _fused_forward_cuda
            cls._auto_fused = True
```

Key design choices:
- **Opt-in via `VLLM_AUTO_FUSE_OPS=1`** — zero behavior change for existing users
- **`dynamic=True`** — handles variable sequence lengths in prefill
- **`mode="default"`** — avoids CUDAGraph capture inside `torch.compile`, which conflicts with vLLM's own graph management
- **Guard `_auto_fused`** — prevents double-patching on re-import

## Benchmark Results

Gemma3-4B-IT, vLLM v0.20.0, NVIDIA RTX PRO 6000 Blackwell (SM120), `enforce_eager=True`.  
Patched classes (auto-detected): `GemmaRMSNorm`, `ReLUSquaredActivation`, `GELU`, `GeluAndMulSparse`, and 9 others (13 total).

### Kernel profile — primary evidence

Single forward pass profiled with `torch.profiler`, prompt_len=1024:

| Metric | Baseline | With fix | Delta |
|---|---|---|---|
| GPU kernel time | 5.064s | 4.343s | **−14.2%** |
| Unfused elementwise events | 115,900 | ~11,000 | **−90%** |
| Triton fused events | 0 | 52,000 | — |
| Unfused % of GPU time | ~40% | <5% | — |

This directly measures the kernel fragmentation problem: 90% of unfused elementwise kernels are eliminated and replaced with fused Triton kernels.

### Decode throughput

decode_len=256, n=15, `enforce_eager=True`:

| batch_size | Baseline TPS | With fix TPS | Delta |
|---|---|---|---|
| 1 | 35.0 | 36.8 | **+5.1%** |
| 4 | 139.5 | 139.6 | ~0% |
| 8 | 277.6 | 277.9 | ~0% |

At bs>1, the decode bottleneck shifts to memory bandwidth (attention), which this patch does not address. The +5% at bs=1 reflects reduced kernel-launch overhead per decode step.

### TTFT (Time to First Token)

End-to-end TTFT via vLLM's synchronous API is confounded by async scheduling jitter; precise attribution requires `AsyncLLMEngine`. The kernel profile above (−14% GPU time, −90% unfused events at prompt_len=1024) is the cleaner signal. Qualitatively, TTFT benefits proportionally to the fraction of prefill time spent in unfused ops; that fraction is ~40% of GPU time in the baseline.

### Memory overhead

Zero additional GPU memory. Compiled artifacts cache to `~/.cache/vllm/torch_compile_cache`.

### Negative control (Llama-3-8B)

No trivial fallback classes; scanner identifies 0 to patch; performance unchanged (delta < 2% noise floor).

## Limitations

1. **First-request compile spike**: ~1.37s on Gemma3-4B. Pre-compiling during `capture_model()` would eliminate this.

2. **No effect under default `VLLM_COMPILE` mode**: When `compilation_config.mode != NONE`, `forward_cuda` is bypassed entirely. This patch only matters for `enforce_eager=True`.

3. **Scope — Class 2 only**: Patches ops where `forward_cuda` is a trivial one-liner. Ops with dtype-mismatch fallbacks (Class 1) or conditional dispatch (Class 3) are out of scope. `torch.compile` does fuse dtype casts as a side effect, so some Class 1 ops get partial improvement.

## Testing

```bash
# Verify patch activates on GemmaRMSNorm
VLLM_AUTO_FUSE_OPS=1 python -c "
from vllm import LLM
llm = LLM('google/gemma-3-4b-it', enforce_eager=True)
from vllm.model_executor.layers.layernorm import GemmaRMSNorm
assert getattr(GemmaRMSNorm, '_auto_fused', False), 'patch not applied'
print('OK:', GemmaRMSNorm._auto_fused)
"

# Negative: no patch without env var
python -c "
from vllm.model_executor.layers.layernorm import GemmaRMSNorm
assert not getattr(GemmaRMSNorm, '_auto_fused', False)
print('OK: no patch')
"
```

`tests/kernels/test_layernorm.py` passes unchanged (tests `forward_native` directly).

## Related

- [vLLM CustomOp abstraction](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/custom_op.py)
- External tool with full taxonomy: [FLARE](https://github.com/[your-repo]/flare) *(preprint forthcoming)*

---

## Notes

**Duplicate check**: searched open PRs for "CustomOp forward_cuda forward_native compile" and "VLLM_AUTO_FUSE_OPS" — no existing PR addresses this pattern.

**AI assistance**: This PR was developed with AI assistance (Claude). All changed lines have been reviewed by the human submitter, benchmark experiments were run on real hardware (NVIDIA RTX PRO 6000 Blackwell, SM120), and the submitter understands and can defend the change end-to-end.

**Tests run**:
```bash
# Verify patch activates on GemmaRMSNorm
VLLM_AUTO_FUSE_OPS=1 python -c "
from vllm import LLM
llm = LLM('google/gemma-3-4b-it', enforce_eager=True)
from vllm.model_executor.layers.layernorm import GemmaRMSNorm
assert getattr(GemmaRMSNorm, '_auto_fused', False)
print('OK')
"
# Negative control: no patch without env var — confirmed
# Llama-3-8B: 0 classes patched, performance unchanged — confirmed
```

### CODE REVIEW

File: vllm/model_executor/custom_op.py [hunk 1]
```diff
@@ -204,6 +240,31 @@ def dispatch_forward(self, compile_native: bool):
         elif current_platform.is_out_of_tree():
             return self.forward_oot
         else:
+            # Auto-fuse trivial forward_cuda fallbacks when opted in.
+            # When VLLM_AUTO_FUSE_OPS=1 and compilation_config.mode == NONE
+            # (enforce_eager=True), ops whose forward_cuda is a one-liner alias
+            # for forward_native produce 6-8 unfused elementwise/reduce kernels
+            # per call. Wrapping forward_native with torch.compile replaces them
+            # with a single fused Triton kernel, reducing GPU kernel time by ~14%.
+            if (
```
[the remark this answers is not in the document]
Comment (Ray-RP) [reply vllm/model_executor/custom_op.py]: you're right. forward_hip defaults to return self.forward_cuda(...) for most ops, so the fragmentation pattern exists equally on ROCm and the previous placement silently skipped it.
Fixed in 335c924: moved the auto-fuse block before the platform if/elif chain and switched to patching forward_native instead of forward_cuda. Since both forward_cuda and forward_hip delegate to forward_native for trivial-fallback ops, this covers all platforms in one place without needing per-platform branches.

File: vllm/model_executor/custom_op.py [hunk 2]
```diff
@@ -204,6 +240,31 @@ def dispatch_forward(self, compile_native: bool):
...
+                os.environ.get("VLLM_AUTO_FUSE_OPS") == "1"
+                and _is_trivial_forward_cuda(self.__class__)
+                and not getattr(self.__class__, "_auto_fused", False)
+            ):
```
[the remark this answers is not in the document]
Comment (Ray-RP) [reply vllm/model_executor/custom_op.py]: The old guard and not getattr(cls, "_auto_fused", False) evaluated _is_trivial_forward_cuda first because Python's and is left-to-right — the guard was on the right and was never reached for non-trivial classes, meaning every instance paid the inspect.getsource cost.
Fixed in 335c924 using the three-state scheme you suggested: _auto_fused = None (unchecked), False (checked, non-trivial — skip cheaply), True (fused). The expensive source inspection now runs at most once per class regardless of how many layer instances exist.

File: vllm/model_executor/custom_op.py [hunk 3]
```diff
@@ -204,6 +240,31 @@ def dispatch_forward(self, compile_native: bool):
...
+                if compilation_config.mode == CompilationMode.NONE:
+                    _original_native = self.__class__.forward_native
+                    # dynamic=True: handles variable sequence lengths (prefill).
+                    # mode="default": avoids CUDAGraph inside torch.compile,
+                    #   which conflicts with vLLM's own CUDAGraph management.
+                    _compiled = torch.compile(
+                        _original_native, dynamic=True, mode="default"
+                    )
+                    def _fused_forward_cuda(self, *args, **kwargs):
+                        return _compiled(self, *args, **kwargs)
```
[the remark this answers is not in the document]
Comment (Ray-RP) [reply vllm/model_executor/custom_op.py]: Good question — we verified this empirically on PyTorch 2.11 (the vLLM v0.20 dependency). When torch.compile wraps an unbound class method, Dynamo guards on type(self), not id(self). Calling the compiled function with different instances of the same class does not trigger recompilation:

fn = GemmaRMSNorm.forward_native   # unbound method
compiled = torch.compile(fn, dynamic=True, mode="default")

### Three different instances, same class
layer1, layer2, layer3 = GemmaRMSNorm(...), GemmaRMSNorm(...), GemmaRMSNorm(...)

compiled(layer1, x)  # calls_captured: 1  ← compile happens here
compiled(layer2, x)  # calls_captured: 1  ← cache hit
compiled(layer3, x)  # calls_captured: 1  ← cache hit

torch._dynamo.utils.counters["stats"]["calls_captured"] stays at 1 across all three calls. This is the expected Dynamo behavior: the weight tensors (self.weight) are traced as symbolic tensors, not bound to a specific instance.

The measured first-request spike is ~295 ms total for Gemma3-4B (34+27 = 61 layers, ~13 trivial classes). That's one compile per class, not one per layer — consistent with the calls_captured result above. The PR description notes this as a known limitation ("~1.37 s on Gemma3-4B") which amortizes after the first request.

**Review by acyngel — APPROVED** [review]

**Comment by Ray-RP** [comment]
Thank you for the review! Could you please add the ready label to trigger CI? 🙏
